scriptworker/cot/verify.py

```diff
@@ -0,0 +1,1141 @@
+#!/usr/bin/env python
+"""Chain of Trust artifact verification.
+
+Attributes:
+    log (logging.Logger): the log object for this module.
+"""
+import asyncio
+from contextlib import contextmanager
+from copy import deepcopy
+from frozendict import frozendict
+import logging
+import os
+import pprint
+import re
+import shlex
+from urllib.parse import unquote, urljoin, urlparse
+from scriptworker.constants import DEFAULT_CONFIG
+from scriptworker.exceptions import CoTError, ScriptWorkerGPGException
+from scriptworker.gpg import get_body, GPG
+from scriptworker.task import download_artifacts, get_decision_task_id, get_worker_type, get_task_id
+from scriptworker.utils import format_json, get_hash, load_json, makedirs, match_url_regex, raise_future_exceptions
+from taskcluster.exceptions import TaskclusterFailure
+
+log = logging.getLogger(__name__)
+
+
+# ChainOfTrust {{{1
+class ChainOfTrust(object):
+    """The master Chain of Trust, tracking all the various `LinkOfTrust`s.
+
+    Attributes:
+        context (scriptworker.context.Context): the scriptworker context
+        cot_dir (str): the local path containing this link's artifacts
+        decision_task_id (str): the task_id of self.task's decision task
+        links (list): the list of `LinkOfTrust`s
+        name (str): the name of the task (e.g., signing.decision)
+        task_id (str): the taskId of the task
+        task_type (str): the task type of the task (e.g., decision, build)
+        worker_impl (str): the taskcluster worker class (e.g., docker-worker) of the task
+    """
+    def __init__(self, context, name, task_id=None):
+        self.name = name
+        self.task_type = guess_task_type(name)
+        self.context = context
+        self.task_id = task_id or get_task_id(context.claim_task)
+        self.task = context.task
+        self.worker_impl = guess_worker_impl(self)  # this should be scriptworker
+        self.decision_task_id = get_decision_task_id(self.task)
+        self.links = []
+
+    def dependent_task_ids(self):
+        """Helper method to get all `task_id`s for all `LinkOfTrust` tasks.
+
+        Returns:
+            list: the list of `task_id`s
+        """
+        return [x.task_id for x in self.links]
+
+    def all_tasks(self):
+        """The list of task definitions for all `self.links` + `self`
+
+        Returns:
+            list of dicts: the task definitions for all links + self
+        """
+        return [self.task] + [x.task for x in self.links]
+
+    def is_try(self):
+        """Helper method to determine if any task in the chain is a try task.
```

Comment:
Yup.  For signing, I have:

* release signing is restricted to [these repos](https://github.com/escapewindow/scriptworker/blob/8e9c08381e049fed24a111da1819f68e3f5e1c78/scriptworker/constants.py#L207-L214)
* nightly signing is restricted to [this superset of repos](https://github.com/escapewindow/scriptworker/blob/8e9c08381e049fed24a111da1819f68e3f5e1c78/scriptworker/constants.py#L220-L230)
* all other repos are able to use dep signing, including try.  That allows us to test the nightly graph on try, even with tasks that depend on signed binaries.  I'm going to push for the other *scripts to follow the same scope pattern, e.g. beetmover won't push into the release bucket from non-release repos.